### DISCUSSION TITLE
allowedVersions doesn't use normalized version

### DISCUSSION THREAD

**Discussion by satanTime** [renovatebot/renovate#25710] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

Placeholder value, please select the correct response from the dropdown

### Wanted end result.

I want to apply `allowedVersions` to `nvm`:

```json
    {
      "matchPackageNames": [
        "satantime/puppeteer-node",
        "@types/node",
        "nodejs/node"
      ],
      "allowedVersions": "/^\\d*[02468]\\./",
      "branchTopic": "node-{{{newMajor}}}{{#if separateMinorPatch}}{{#if isPatch}}.{{{newMinor}}}{{/if}}{{/if}}.x{{#if isLockfileUpdate}}-lockfile{{/if}}"
    }
```

However, all PRs don't have `.nvmrc` at all, despite `nodejs/node`.
Job logs show:

```json
      {
        "deps": [
          {
            "depName": "node",
            "currentValue": "18.18.2",
            "datasource": "github-tags",
            "packageName": "nodejs/node",
            "updates": [],
            "versioning": "node",
            "warnings": [],
            "sourceUrl": "https://github.com/nodejs/node",
            "registryUrl": "https://github.com",
            "currentVersion": "18.18.2",
            "isSingleVersion": true,
            "fixedVersion": "18.18.2"
          }
        ],
        "packageFile": ".nvmrc"
      },
```

It looks obvious that based on `currentValue`, `currentVersion`, `fixedVersion` it should be updated to `20.x.x`, but nothing happens.

The issue here is that `nvm` version includes `v`, and the right `allowedVersions` is `^v?\\d*[02468]\\./"`, but it's impossible to guess about this.

I would propose to apply `allowedVersions` to the parsed (extracted) version, so I wouldn't worry whether `v` should be or shouldn't be here.

### What you tried so far.

checked logs to find a way to identify how I should discover `v` - but logs don't show anything about this.

### Relevant debug logs

<details><summary>Logs</summary>

```
      {
        "deps": [
          {
            "depName": "node",
            "currentValue": "18.18.2",
            "datasource": "github-tags",
            "packageName": "nodejs/node",
            "updates": [],
            "versioning": "node",
            "warnings": [],
            "sourceUrl": "https://github.com/nodejs/node",
            "registryUrl": "https://github.com",
            "currentVersion": "18.18.2",
            "isSingleVersion": true,
            "fixedVersion": "18.18.2"
          }
        ],
        "packageFile": ".nvmrc"
      },
```

</details>

**Comment by rarkins** [comment]
Assuming you are trying to do this for Node LTS reasons, are you aware of `node` versioning?

**Reply by satanTime** [reply]
node here is just for instance, the main issue is that `nvm` versions itself with `v`, whereas `node` and `npm` don't.

so the goal here was to create a single pattern for versions among different repos / packages, but this `v` prefix in `nvm` breaks the approach, and based on renovate logs it's not clear why `nvm` fails with `allowedVersions`.

**Reply by rarkins** [reply]
And doesn't the `v?` as optional v work?

**Reply by satanTime** [reply]
with `v?` it works, but it wasn't clear based on logs. So I removed everything from my config and started to add fields one by one to find what causes issues with `nvm`.

Based on docs: https://docs.renovatebot.com/templates/, it looks like the version is already normalized, there is not a variable which would contain `v`.

Also based on docs: https://docs.renovatebot.com/configuration-options/#extractversion, the impression was that the very first phase is to extract the current version by `extractVersion`, and the result of this operation is used for `allowedVersions`, `newMajor`, `matchCurrentVersion` etc.

For example, I can easily imagine a special docker image I need with node and nvm which version is `stable-node-20.0.0`.
In this case, `allowedVersions` should be `"/^(v|stable-node-)?\\d*[02468]\\./"` what is more confusing and isn't easily guessable.

**Reply by satanTime** [reply]
One more note - in the docs here: https://docs.renovatebot.com/configuration-options/#allowedversions
it shows an example of `"allowedVersions": "<=1.5"`, but would it really work if the raw version is `stable-node-1.5.0`? or would it stumble on parsing `stable-node-` and fail?

**Reply by rarkins** [reply]
Can you create a reproduction of one of the use cases which doesn't work as you'd expect? I'd like to debug it

**Reply by satanTime** [reply]
sure, I'll do it tomorrow or day after, and post here.

**Reply by satanTime** [reply]
Hi there, happy Thursday.

I'm trying to recreate a min repo, however, I cannot get `stable-verion-` working.

https://developer.mend.io/github/satanTime/test2/-/job/018bd578-2933-787c-9536-724bfdc03572
```
DEBUG: Found no results from datasource that look like a version
{
  "baseBranch": "main"
  "dependency": "satantime/test"
  "result": {
    "registryUrl": "https://index.docker.io",
    "releases": []
  }
}
```

Whereas https://hub.docker.com/r/satantime/test/tags has several versions.
